src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala

```diff
@@ -180,17 +182,21 @@ private[health] class HealthCheckActor(
             case None =>
               log.error(s"Couldn't find instance $instanceId")
               health.update(result)
-          }
+          })
       }
 
-      healthByInstanceId += (instanceId -> newHealth)
+      futureHealth.onComplete {
+        case Success(newHealth) =>
+          healthByInstanceId += (instanceId -> newHealth)
```

Comment:
Agree, this is dangerous. How about sending complete function back to actor to have proper locking and state updating? Actor will have another case class (e.g., Complete) that will contain function that need to be run in order to complete job.